### PR TITLE
project config: dashboard-cephadm-e2e

### PR DESCRIPTION
**Pull request by alfonsomthd**
To be merged after https://github.com/ceph/ceph/pull/42263

Project: run Dashboard E2E tests with cephadm as orchestrator backend
  on dashboard/cephadm PRs.

Fixes: https://tracker.ceph.com/issues/51301

Signed-off-by: Alfonso Martínez <almartin@redhat.com>

### CODE REVIEW

File: scripts/dashboard/install-kcli-deps.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+echo "
+alias kcli='podman run --rm --net host --security-opt label=disable \
+    -v ${HOME}/.ssh:/root/.ssh \
+    -v ${HOME}/.kcli:/root/.kcli \
+    -v ${PWD}/src/pybind/mgr/dashboard/ci/cephadm:/workdir \
+    -v /var/lib/libvirt/images:/var/lib/libvirt/images \
+    -v /var/run/libvirt:/var/run/libvirt \
+    -v /var/tmp:/ignitiondir \
+    docker.io/jolmomar/kcli'
```
Comment (jmolmo) [inline scripts/dashboard/install-kcli-deps.sh:23]: I think it could be interestiong to use a constant to "easy" change the kcli image... one day we will use latest ... (i hope)

File: scripts/dashboard/install-kcli-deps.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+newgrp libvirt
+systemctl enable --now libvirtd
+
+sudo dnf install -y podman
+
+mkdir -p ~/.ssh
+cp -R src/pybind/mgr/dashboard/ci/cephadm/.kcli ~
+
+echo "
+alias kcli='podman run --rm --net host --security-opt label=disable \
```
Comment (jmolmo) [inline scripts/dashboard/install-kcli-deps.sh:16]: what about the  "--it" parameter?

File: scripts/dashboard/install-kcli-deps.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    -v ${HOME}/.ssh:/root/.ssh \
+    -v ${HOME}/.kcli:/root/.kcli \
+    -v ${PWD}/src/pybind/mgr/dashboard/ci/cephadm:/workdir \
+    -v /var/lib/libvirt/images:/var/lib/libvirt/images \
+    -v /var/run/libvirt:/var/run/libvirt \
+    -v /var/tmp:/ignitiondir \
```
Comment (jmolmo) [inline scripts/dashboard/install-kcli-deps.sh:22]: I think you forget: 
`-v $PWD:/workdir `

**Review by jmolmo — APPROVED** [review]
Minor nits with the kcli command... apart of that ... fantastic job Alfonso!!

File: scripts/dashboard/install-kcli-deps.sh (hunk 2 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-kcli-deps.sh:16]: With that I cannot capture the output that I need later on:
```
if [[ -z "${CYPRESS_BASE_URL}" ]]; then
    CYPRESS_BASE_URL="https://$(kcli info vm ceph-node-00 -f ip -v):8443"
fi
```

File: scripts/dashboard/install-kcli-deps.sh (hunk 3 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-kcli-deps.sh:22]: In fact is set to:
-v ${PWD}/src/pybind/mgr/dashboard/ci/cephadm:/workdir

File: scripts/dashboard/install-kcli-deps.sh (hunk 1 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-kcli-deps.sh:23]: Yes, for now I'm testing the job. I'll replace it with an env. var.

File: scripts/dashboard/install-kcli-deps.sh (hunk 3 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-kcli-deps.sh]: Updated anyway.

File: ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    triggers:
+      - github-pull-request:
+          cancel-builds-on-update: true
+          allow-whitelist-orgs-as-admins: true
+          org-list:
+            - ceph
+          white-list-labels:
+#            - dashboard
+            - wip-alfonsomthd-testing
```
Comment (epuertat) [inline ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml:39]: Update?

File: ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+          black-list-target-branches:
+            - luminous
+            - mimic
+            - nautilus
+            - octopus
```
Comment (epuertat) [inline ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml:44]: Shouldn't we run this on octopus ones too?

File: ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    project-type: freestyle
+    defaults: global
+    concurrent: true
+    node: huge && focal && x86_64
+    display-name: 'ceph: Dashboard + Cephadm E2E'
+    quiet-period: 5
+    block-downstream: false
+    block-upstream: false
+    retry-count: 3
+    properties:
```
Comment (epuertat) [inline ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml:12]: As discussed with you about avoiding concurrent jobs in the same node, we'd need the [Throttle plugin](https://plugins.jenkins.io/throttle-concurrents/) installed and [configured in JJB](https://docs.openstack.org/infra/jenkins-job-builder/properties.html?highlight=throttle#properties.throttle).

**Review by epuertat — APPROVED** [review]
Looks good! Thanks @alfonsomthd ! Just an observation about the throttling.

File: ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml (hunk 6 above)
Comment (alfonsomthd) [reply ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml:12]: @djgalloway Can you help us here?

Comment (djgalloway) [reply ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml:12]: `concurrent` means "at the same time anywhere."  The number of workers/executors per Jenkins builder determines how many jobs can run on a node at a time.

tl;dr - you want `concurrent: true`

File: scripts/dashboard/install-cephadm-e2e-deps.sh [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+sudo systemctl enable --now libvirtd
+
+# Add podman repo for ubuntu focal.
+VERSION_ID='20.04'
+sudo sh -c "echo 'deb http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_${VERSION_ID}/ /' > /etc/apt/sources.list.d/devel:kubic:libcontainers:stable.list"
+wget -nv https://download.opensuse.org/repositories/devel:kubic:libcontainers:stable/xUbuntu_${VERSION_ID}/Release.key -O- | sudo apt-key add -
+sudo apt update -y
+
+# Install required deps.
+sudo apt install -y nodejs npm openssh-server podman
```
Comment (djgalloway) [inline scripts/dashboard/install-cephadm-e2e-deps.sh:31]: I foresee all of this being very problematic.  To be honest, I'm a little disheartened to see it's already been run against some of the Jenkins builders as it will no doubt create headaches in the future.

These aren't ephemeral builders.  I do my best to make sure all the builders are running the same OS/kernel/packages/etc.  All of this prep is bound to eventually cause package conflicts and `apt` transaction errors when opensuse's servers are down, etc.

File: scripts/dashboard/install-cephadm-e2e-deps.sh [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+mkdir -p ~/.kcli ~/.ssh
+if [[ ! -f "$HOME/.ssh/id_rsa" ]]; then
+    ssh-keygen -t rsa -q -f "$HOME/.ssh/id_rsa" -N ""
```
Comment (djgalloway) [inline scripts/dashboard/install-cephadm-e2e-deps.sh:35]: Please name it something unique to this test

Comment (djgalloway) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:35]: Maybe symlink to it inside the container temporarily or something.

File: scripts/dashboard/install-cephadm-e2e-deps.sh [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    -v /var/run/libvirt:/var/run/libvirt \
+    -v /var/tmp:/ignitiondir \
+    ${KCLI_CONTAINER_IMAGE} \""'${@}'"\"
+" | sudo tee /usr/local/bin/kcli
+sudo chmod +x /usr/local/bin/kcli
+
+podman info
+sudo podman container prune -f
+
+kcli create pool -p /var/lib/libvirt/images default
```
Comment (djgalloway) [inline scripts/dashboard/install-cephadm-e2e-deps.sh:57]: Does this ever get cleaned up?

File: scripts/dashboard/install-cephadm-e2e-deps.sh [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+kcli download image fedora34 -u https://fedora.mirror.liteserver.nl/linux/releases/34/Cloud/x86_64/images/Fedora-Cloud-Base-34-1.2.x86_64.qcow2
+kcli create network -c 192.168.122.0/24 default
```
Comment (djgalloway) [inline scripts/dashboard/install-cephadm-e2e-deps.sh:59]: Same question here.

File: scripts/dashboard/install-cephadm-e2e-deps.sh (hunk 8 above)
Comment (djgalloway) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:35]: Or better yet, write the key to `$WORKSPACE/id_rsa` and mount that inside the container at `~/.ssh`

**Comment by djgalloway** [comment]
This all looks great in theory but it modifies the Jenkins builders and relies on third-party infrastructure a bit more than I would like.

I would rather have a dedicated builder or few builders that have libvirt/podman stuff installed and give them a new Jenkins label (`node:` in the jjb yaml).  What do you think?

File: ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml (hunk 6 above)
Comment (epuertat) [reply ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml:12]: @djgalloway just the opposite: as per @alfonsomthd words, more than a single `kcli` instance running might cause errors. That Throttling plugin allows to set the number of concurrent jobs per node.

File: scripts/dashboard/install-cephadm-e2e-deps.sh (hunk 7 above)
Comment (epuertat) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:31]: @alfonsomthd tried to avoid docker for the root/daemon security issue. I think it's probably more standard/straightforward to rely on root/daemon docker than to have to install podman from an non-standard source.

@djgalloway we can label these nodes as 'cephadm-kcli' (and use that label in this very job) so we have a precise control over which hosts this is run. And for avoiding installing packages during the testing process in the future, what about labeling a node or nodes as 'sandbox' so we have some 'damage control'.

Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:31]: > I foresee all of this being very problematic. To be honest, I'm a little disheartened to see it's already been run against some of the Jenkins builders as it will no doubt create headaches in the future.
> 

I ran it as a PoC, is it problematic testing the job to check if the idea is feasible?

> These aren't ephemeral builders. I do my best to make sure all the builders are running the same OS/kernel/packages/etc. All of this prep is bound to eventually cause package conflicts and `apt` transaction errors when opensuse's servers are down, etc.

I chose podman toavoid installing docker as a sustemd service (for simplicity).
I rely on that build because ubuntu focal does not have yet podman available as deb pkg.

File: scripts/dashboard/install-cephadm-e2e-deps.sh (hunk 8 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:35]: The keys are not only used by podman but by KCLI, the VM manager, so the VMs are using it so form the host you can SSH to any VM. I tried with a dedicated key and I was not able to SSH into any VM.
Nevertheless I'll investigate this a bit more.

File: scripts/dashboard/install-cephadm-e2e-deps.sh (hunk 9 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:57]: I do some cleanup in the associated ceph script but it does not include this.

Following your idea of having dedicated builders with libvirt+podman(or docker) and taking into account that the VM image is never modified, not deleting the image can improve the speed of the job.
Does it make sense to not do the cleanup in the dedicated nodes?

**Comment by alfonsomthd** [comment]
> This all looks great in theory but it modifies the Jenkins builders and relies on third-party infrastructure a bit more than I would like.
> 
> I would rather have a dedicated builder or few builders that have libvirt/podman stuff installed and give them a new Jenkins label (`node:` in the jjb yaml). What do you think?

Sounds good to me!
So I assume that the job shouldn't install anymore these tools and only take care of setting the proper environment and launch the job script.

File: ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml (hunk 4 above)
Comment (alfonsomthd) [reply ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml:39]: This will be the last step before merging XD

File: ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml (hunk 5 above)
Comment (alfonsomthd) [reply ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml]: As per our previous conversation will run it in octopus but take into account that the octopus ceph backport will be needed to succeed.

File: scripts/dashboard/install-cephadm-e2e-deps.sh (hunk 8 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh]: @djgalloway Now we store the key in KCLi config folder:
```
KCLI_CONFIG_DIR="${HOME}/.kcli"
mkdir -p ${KCLI_CONFIG_DIR}
if [[ ! -f "${KCLI_CONFIG_DIR}/id_rsa" ]]; then
    ssh-keygen -t rsa -q -f "${KCLI_CONFIG_DIR}/id_rsa" -N ""
fi
```
Let us know if this works for you.

File: ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml (hunk 6 above)
Comment (alfonsomthd) [reply ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml:12]: @djgalloway As per our previous conversations we concluded that putting `concurrent: true` does not imply that more than 1 execution of the same job can run on the same machine, so it is safe to set it as true.

File: scripts/dashboard/install-cephadm-e2e-deps.sh (hunk 7 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:31]: @djgalloway As per our previous conversation we agreed on using docker instead of podman.

File: scripts/dashboard/install-cephadm-e2e-deps.sh (hunk 9 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:56]: @djgalloway As per our last conversation we agreed on using the `node: huge && focal && x86_64` machines and a dedicated libvirt pool and network (ceph-dashboard) to avoid conflicts in the future.

File: scripts/dashboard/install-cephadm-e2e-deps.sh (hunk 10 above)
Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:58]: Answered above.

File: scripts/dashboard/install-cephadm-e2e-deps.sh [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+" | sudo tee /usr/local/bin/kcli
+sudo chmod +x /usr/local/bin/kcli
+
+# Install required deps.
+sudo apt update -y
+sudo apt install -y nodejs npm openssh-server
+
+sudo mkdir -p /var/lib/libvirt/images/ceph-dashboard
+kcli create pool -p /var/lib/libvirt/images/ceph-dashboard ceph-dashboard
+kcli create network -c 192.168.100.0/24 ceph-dashboard
```
Comment (djgalloway) [inline scripts/dashboard/install-cephadm-e2e-deps.sh:74]: Can you add a teardown script to clean this up after the job completes?

Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:74]: @djgalloway The cleanup is tirggered when the job finishes on success/error here (in the other PR):
https://github.com/ceph/ceph/pull/42263/files?file-filters%5B%5D=.sh#diff-ccaf2a6af4475351e29cd34ed246c9fbd3cbbc14c93d961a360c283eecf02e88R5

In our last conversation we agreed to not erase the libvirt folder/pool/network but to rename it (s/default/ceph-dashboard) in order to not constantly download the VM fedora image and to avoid name conflicts in the future.
Do you prefer to completely erase it?

Comment (djgalloway) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:74]: Ah, right.  Fair enough.  Can you just put a comment in the file pointing to that cleanup function please?  The canonical link to the file in `master` is fine even though that PR isn't merged yet.

Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:74]: Sure.

Comment (alfonsomthd) [reply scripts/dashboard/install-cephadm-e2e-deps.sh:75]: Done.

**Review by djgalloway — APPROVED** [review]
Can we wait to merge until Monday please?  I'm on PTO tomorrow and won't be available to deal with the fallout if things go bad.

**Comment by alfonsomthd** [comment]
> Can we wait to merge until Monday please? I'm on PTO tomorrow and won't be available to deal with the fallout if things go bad.

Sure. We'll merge https://github.com/ceph/ceph/pull/42263 first.
